### PR TITLE
#1: initial implementation of the mini-app

### LINKED ISSUE #1: Create mesh class and test it
The goal of this issue is to translate the python mesh class into c++ and to ensure that it properly functions.

### PR DESCRIPTION
**Pull request by pierrepebay**
Fixes #1 ; this is the implementation of the mesh class from the python code

### CODE REVIEW

File: cppfd/CMakeLists.txt
Comment (lifflander) [inline cppfd/CMakeLists.txt:8]: This is ok as a hack, but you shouldn't have to do this.

Comment (pierrepebay) [reply cppfd/CMakeLists.txt:8]: Thanks @lifflander, but I cannot get the clang compiler to look at this directory for includes. Is this a CMake variable that I should add at configuration time instead?

File: cppfd/boundaryconditions.cpp
Comment (ppebay) [inline cppfd/boundaryconditions.cpp]: I would rename these member variables (`u_<location>` and `v_<location>`) to make them less confusing as `v` might be confused with the velocity itself, while it only denotes its second spatial component. 
Suggestion: `velocity_x_top` for `u_top` and `velocity_y_top` for `v_top` (and similarly for other locations)>

File: cppfd/solver.cpp
Comment (ppebay) [inline cppfd/solver.cpp:19]: Add comment here to explain that this is the computation of the Reynolds number (not a primary input to the code)

File: cppfd/solver.cpp
Comment (ppebay) [inline cppfd/solver.cpp:68]: Add comment here to indicate that the flat index of a cell into the Cartesian mesh becomes the per-dimension Cartesian index into the Laplacian matrix.

**Review by ppebay — COMMENTED** [review]
Not approving yet as implementation not complete yet -- however @pierrepebay the comments of this partial review can already be implemented.

File: cppfd/CMakeLists.txt
Comment (ppebay) [reply cppfd/CMakeLists.txt:9]: @lifflander any idea why this seems to be needed to get the code to compile? I tried to remove it but it failed.

File: cppfd/boundaryconditions.cpp
Comment (ppebay) [reply cppfd/boundaryconditions.cpp]: Making this a separate issue for later.

**Comment by ppebay** [comment]
Reviewed entire code and it appears sound, numerically. 

Needs to be tested now but approving merge.

**Review by ppebay — APPROVED** [review]

File: cppfd/solver.cpp
Comment (lifflander) [inline cppfd/solver.cpp:268]: Extra white space here.

**Review by lifflander — CHANGES_REQUESTED** [review]
I found a few things that could use improvement before merging. In general, there are some stylistic things that don't follow our [C++ style guideline](https://github.com/DARMA-tasking/vt/wiki/Code-Style-Guidelines). You don't have to exactly follow it because this is not in VT, but I recommend a few changes to make it easier to read.

File: cppfd/CMakeLists.txt
Comment (lifflander) [reply cppfd/CMakeLists.txt:9]: When I remove this on my system, it works perfectly fine. Can you show me the exact error you are getting?

File: cppfd/mesh.h
Comment (ppebay) [reply cppfd/mesh.h:16]: @lifflander : regarding the second point (curly braces vs. parens for initialization), what would be the preferred style in the context of DARMA? And what would be the rationale? Thanks!

File: cppfd/CMakeLists.txt
Comment (pierrepebay) [reply cppfd/CMakeLists.txt:9]: It is now working without this line, it might have to do with how I installed Kokkos.

File: cppfd/CMakeLists.txt
Comment (pierrepebay) [reply cppfd/CMakeLists.txt:18]: When I started creating the classes it was useful to verify the different functionalities. Also, in the context of testing, wouldn't it be useful to be able to use the libraries independently ?

Comment (ppebay) [reply cppfd/CMakeLists.txt:17]: Yes, I think at this point it's useful to test them independently.

File: cppfd/mesh.h
Comment (lifflander) [reply cppfd/mesh.h]: For initializing constructors, we typically prefer parenthesis unless curly braces are needed due to semantic ambiguity (most vexing parse). Curly braces can also be used to list initialization so there can be ambiguity there also. For initializing constructors, it's extremely rare to need curly braces (in fact, there are zero cases in VT where it's required).

For regular variable initialization/declaration, it happens far more often that you need curly braces so we commonly use that instead parenthesis---in fact, it is preferred.

Comment (ppebay) [reply cppfd/mesh.h]: Noted, thank you! This is very useful.
I don't recall having seen these recs into the coding style guidelines, would be useful to add them if they are not.

**Review by ppebay — APPROVED** [review]

**Review by lifflander — APPROVED** [review]
Looks good to me, ready to merge once you get the signatures fixed.

**Pull request merged**
